Interposer tag

AmirhosseinPoolad opened the pull request:
This PR contains an initial proposal for a new interposer tag. To define an interposer cut, you would add an <interposer_cut dim=x|y loc=val> tag under a `<layout>` tag that defines the location of the cut and the axis of the cut. This tag would then have multiple <interdie_wire> children specifying the connections between the two sides of the cut-line.

> <interdie_wire sg_name="interposer_conn_sg" sg_link="X4" offset_start="0" offset_end="2" offset_increment="1" num=50/>

- sg_name: name of the scatter-gather pattern to be used for the connections
- sg_link: name of the sg_link to be used for the connections
- offset_start,end,increment: Connections can be laid out in various offsets from the cut-line, these options
- num: number of scatter-gather patterns to be instantiated at each switchblock location

Example:

<img width="731" height="507" alt="image" src="https://github.com/user-attachments/assets/f76b3dad-a985-4c3c-8d9f-ea1978ad2503" />

Note that in this example we have 3 different switchblock locations and 3*50=150 sg patterns would be instantiated.

AmirhosseinPoolad commented:
I think the code is ready for review now! @vaughnbetz @soheilshahrouz

vaughnbetz commented on doc/src/arch/reference.rst:
```diff
@@ -581,6 +581,32 @@ Grid Layout Example
+.. arch:tag:: <interposer_cut dim=x|y loc="int"/>
+
+    :req_param dim: Dimension or axis of the cut. 'X' or 'x' means a horizontal cut while 'Y' or 'y' means a vertical cut.
+    :req_param loc: Location of the cut. Currently only absolute positions are supported.
```
Can this be an equation, or does it have to be a number?  Should say ...

vaughnbetz replied:
Or maybe that's what absolute positions means?

AmirhosseinPoolad replied:
Yes, it was absolute versus relative [to device height/width] position, as in numbers only. Changed the comment to say "integer values" instead.

vaughnbetz commented on doc/src/arch/reference.rst:
```diff
@@ -581,6 +581,32 @@ Grid Layout Example
+    .. figure:: scatter_gather_images/interposer_diagram.png
+        
+        An example of how specifying interposers in VTR works. Connections between the two sides of a cut are first severed. The two sides are then reconnected using scatter_gather patterns. Note that there are 'num' of each pattern at each switchblock locations.
+
```
1. I think the figure should show the die-crossing wires at each x location, since that's what would happen.
2. On the figure or in the text, say the length of the sg_link wire in this case is 3.

vaughnbetz replied:
I assume the interposer cuts all layers?  Should say that too.  Or say it only supports 1 layer architectures for now.

AmirhosseinPoolad replied:
I'm not entirely sure about what happens with 3D architectures and interposers. The idea I had was that each layer could have it's own independent cut lines to keep everything flexible. This might make physical sense for a 2 layer face-to-face stacked chip, but I'm not sure about that either.

Right now the code does independent cut lines for each layer, but I could change it if you think it wouldn't make physical sense.

vaughnbetz commented on libs/libarchfpga/src/interposer_types.h:
```diff
@@ -0,0 +1,51 @@
+ */
+struct t_interposer_cut_inf {
+    e_interposer_cut_dim dim;                        ///< Dimension or axis of interposer cut
+    int loc;                                         ///< Location of the cut on the grid
```
Give an example for this one.  If the interposer cut is at y = 13, is the cut above or below the tiles with coordinate y = 13?

vaughnbetz commented on libs/libarchfpga/src/interposer_types.h:
```diff
@@ -0,0 +1,51 @@
+struct t_interdie_wire_inf {
+    std::string sg_name;  ///< Name of the scatter-gather pattern to be used for the interdie connection
+    std::string sg_link;  ///< Name of the scatter-gather link to be used for the interdie connection
+    int offset_start;     ///< Starting point of scatter-gather instantiations
```
I think you can beef up these comments a bit.  If the interposer cut is at y = 13, is an offset of 0 for an upward going sg_link a switch block / wire start point at y = 13?  For a downward going wire with a cut at y = 13, is an offset of 0 a sg_link wire start point at y = 14?
Are the start and end points inclusive (I think they are, but you should say that).

vaughnbetz commented on libs/libarchfpga/src/read_xml_arch_file_interposer.cpp:
```diff
@@ -0,0 +1,44 @@
+    interposer.dim = CHAR_INTERPOSER_DIM_MAP.at(interposer_dim[0]);
+
+    interposer.loc = pugiutil::get_attribute(interposer_cut_tag, "loc", loc_data).as_int();
+    if (interposer.loc < 0) {
```
What does it mean if the interposer location is 0?  Goes back to the question of which side of an interposer the specified value goes on.  If the lower side, then a loc of 0 makes sense. Otherwise it doesn't make sense.

vaughnbetz commented on libs/libarchfpga/src/grid_types.h:
```diff
@@ -0,0 +1,131 @@
+ *                            <-------------->
+ *                                 repeatx
+ *
+ * startx/endx and endx/endy define a rectangular region instances dimensions.
```
instances -> instance's

vaughnbetz commented on libs/libarchfpga/src/grid_types.h:
```diff
@@ -0,0 +1,131 @@
+
+    std::string block_type; //The block type name
+
+    int priority = 0; //Priority of the specification.
```
Make these member comments Doxygen format

vaughnbetz commented on libs/libarchfpga/src/grid_types.h:
```diff
@@ -0,0 +1,131 @@
+ * and 'H' (device height). Formulas can be evaluated using parse_formula()
+ * from expr_eval.h.
+ */
+struct t_grid_loc_spec {
```
I think we should build these into the grid part of the VPR API documentation on the web. You'd need to edit some doxyfile for that, and add an overview at the top of this struct saying this is used to capture the user's specification of the architecture will eventually be turned into a flattened grid.

vaughnbetz commented on libs/libarchfpga/src/grid_types.h:
```diff
@@ -0,0 +1,131 @@
+                                     // that come from a common definition.
+};
+
+enum class GridDefType {
```
Add Doxygen comment.

vaughnbetz commented on libs/libarchfpga/src/physical_types.h:
```diff
@@ -248,126 +250,10 @@ typedef enum e_power_estimation_method_ t_power_estimation_method;
 
 struct t_layer_def {
-    std::vector<t_grid_loc_def> loc_defs; //The list of block location definitions for this layer specification
+    std::vector<t_grid_loc_def> loc_defs;              ///< List of block location definitions for this layer specification
```
It would be good to make an overall Doxygen comment for what this struct is intended for.

vaughnbetz reviewed (approved):
I suggest a few more commenting updates, and it would be good to get the grid_def data structure in the online documentation (maybe as another item in the grid tab).

Or, maybe we should put the architecture file capturing structs (grid_def and interposer) in a new Doxygen tab for architecture constructs?

AmirhosseinPoolad commented:
Added the requested changes. For now, I added the doxygen documentation for the grid structs in the grid tab of VPR, but it might be useful to have a LibArchFPGA specific tab. Unfortunately much of LibArchFPGA is not Doxygen documented so it won't be much of a tab.

soheilshahrouz commented:
Is there a test for parsing this new tag?

vaughnbetz commented:
I like the idea of a libarchfpga tab, even if it is currently sparse.  Could you add that?

@soheilshahrouz is also right that you should add a parsing test.

AmirhosseinPoolad commented:
Added a test and also a libarchfpga tab to docs. The added tab is pretty sparse at the moment, but I think fixing that would go beyond the scope of this PR.

vaughnbetz commented on doc/src/api/libarchfpga/physical_types.rst:
```diff
@@ -0,0 +1,33 @@
+Physical Types
+========
+
+These types are used to capture user's intended architecture specification and will be later turned into a flattened device grid according to the device's size.
```
This comment isn't right ; this goes with the grid defs, but not the physical types.

AmirhosseinPoolad replied:
I used the grid defs page as a template for the other pages and forgot to change this one's description. Fixed it now.

vaughnbetz commented on doc/src/api/libarchfpga/scatter_gather_types.rst:
```diff
@@ -0,0 +1,17 @@
+Scatter-Gather Types
+========
+
+These types store information about about scatter-gather patterns
```
about scatter-gather routing patterns (I'd add the word routing).

vaughnbetz reviewed (approved):
Looks good, but I have two minor nits about the high-level doxygen comments. If you fix those, it's good to merge.
